Multiple SortedSet in Redis sink not working

# Issue Guidelines

Please review these questions before submitting any issue?

### What version of the Stream Rector are you reporting this issue for?
1.0.0

### Are you running the correct version of Kafka/Confluent for the Stream reactor release?
Yes, kafka 2.12-1.0.0

### Have you consulted our [FAQs](http://docs.datamountaineer.com/en/latest/faq.html) page first?
Yes, I have.

### Do you have a supported version of the data source/sink .i.e Cassandra 3.0.9?
I hope so but the exception occurs when connect's configuration is parsed before trying to connect to redis.

### Have you read the [docs](http://lenses.stream/connectors/index.html)?
Yes, I have.

### What is the expected behaviour?
Connect to redis instance.

### What was observed?
An exception occurs when a kcql from manual (_SELECT temperature, humidity FROM sensorsTopic PK sensorID STOREAS SortedSet(score=timestamp)_) is used:
```
[2018-05-08 15:05:56,586] ERROR WorkerSinkTask{id=redis-signals-sink-0} Task threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerTask)
java.lang.NullPointerException
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask$$anonfun$4.apply(RedisSinkTask.scala:66)
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask$$anonfun$4.apply(RedisSinkTask.scala:66)
        at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:248)
        at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
        at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:247)
        at scala.collection.TraversableLike$class.filter(TraversableLike.scala:259)
        at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask.start(RedisSinkTask.scala:66)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.initializeAndStart(WorkerSinkTask.java:267)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:163)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:170)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:214)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[2018-05-08 15:05:56,589] ERROR WorkerSinkTask{id=redis-sink-0} Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask)
```
I can run sink in cache and sorted set mode but not in multiple sorted set mode. 
When trying to set a prefix: _INSERT INTO SENSOR- SELECT temperature, humidity FROM sensorsTopic PK sensorID STOREAS SortedSet(score=timestamp)_ there is a different exception:
```
java.lang.AssertionError: assertion failed: They keyword PK (Primary Key) is not supported in Redis INSERT_SS mode. Please review the KCQL syntax of connector
        at scala.Predef$.assert(Predef.scala:170)
        at com.datamountaineer.streamreactor.connect.redis.sink.writer.RedisInsertSortedSet$$anonfun$3.apply(RedisInsertSortedSet.scala:50)
        at com.datamountaineer.streamreactor.connect.redis.sink.writer.RedisInsertSortedSet$$anonfun$3.apply(RedisInsertSortedSet.scala:45)
        at scala.collection.immutable.Set$Set1.foreach(Set.scala:94)
        at com.datamountaineer.streamreactor.connect.redis.sink.writer.RedisInsertSortedSet.<init>(RedisInsertSortedSet.scala:45)
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask$$anonfun$start$2.apply(RedisSinkTask.scala:77)
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask$$anonfun$start$2.apply(RedisSinkTask.scala:75)
        at scala.Option.map(Option.scala:146)
        at com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask.start(RedisSinkTask.scala:75)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.initializeAndStart(WorkerSinkTask.java:267)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:163)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:170)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:214)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```


### What is your Connect cluster configuration (connect-avro-distributed.properties)?
I've tried it in a standalone mode.

### What is your connector properties configuration (my-connector.properties)?
name=redis-sink
connector.class=com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkConnector
tasks.max=1
topics=sensorsTopic
connect.redis.host=127.0.0.1
connect.redis.port=6379
connect.redis.kcql=SELECT temperature, humidity FROM sensorsTopic PK sensorID STOREAS SortedSet(score=timestamp)

--- a/kafka-connect-redis/src/main/scala/com/datamountaineer/streamreactor/connect/redis/sink/RedisSinkTask.scala
+++ b/kafka-connect-redis/src/main/scala/com/datamountaineer/streamreactor/connect/redis/sink/RedisSinkTask.scala
@@ -61,11 +61,24 @@ class RedisSinkTask extends SinkTask with StrictLogging {
     //-- Find out the Connector modes (cache | INSERT (SortedSet) | PK (SortedSetS)
 
     // Cache mode requires >= 1 PK and *NO* STOREAS SortedSet setting
-    val modeCache = settings.copy(kcqlSettings = settings.kcqlSettings.filter(_.kcqlConfig.getStoredAs == null).filter(_.kcqlConfig.getPrimaryKeys.size() >= 1))
+    val modeCache = settings.copy(kcqlSettings =
+      settings.kcqlSettings
+        .filter(k => k.kcqlConfig.getStoredAs == null
+          && k.kcqlConfig.getPrimaryKeys.size() >= 1))
+
     // Insert Sorted Set mode requires: target name of SortedSet to be defined and STOREAS SortedSet syntax to be provided
-    val mode_INSERT_SS = settings.copy(kcqlSettings = settings.kcqlSettings.filter(_.kcqlConfig.getStoredAs == "SortedSet").filter(_.kcqlConfig.getTarget.length > 0))
+    val mode_INSERT_SS = settings.copy(kcqlSettings =
+      settings.kcqlSettings
+      .filter(k => k.kcqlConfig.getStoredAs.toUpperCase == "SORTEDSET"
+        && k.kcqlConfig.getTarget != null
+        && k.kcqlConfig.getPrimaryKeys == null)
+    )
+
     // Multiple Sorted Sets mode requires: 1 Primary Key to be defined and STORE SortedSet syntax to be provided
-    val mode_PK_SS = settings.copy(kcqlSettings = settings.kcqlSettings.filter(_.kcqlConfig.getStoredAs == "SortedSet").filter(_.kcqlConfig.getPrimaryKeys.length == 1))
+    val mode_PK_SS = settings.copy(kcqlSettings =
+      settings.kcqlSettings
+        .filter(k => k.kcqlConfig.getStoredAs.toUpperCase == "SORTEDSET"
+          && k.kcqlConfig.getPrimaryKeys.length == 1))
 
     //-- Start as many writers as required
     writer =
@@ -77,7 +90,7 @@ class RedisSinkTask extends SinkTask with StrictLogging {
         List(new RedisInsertSortedSet(mode_INSERT_SS))
       } ++ mode_PK_SS.kcqlSettings.headOption.map { _ =>
         logger.info("Starting " + mode_PK_SS.kcqlSettings.size + " KCQLs with Redis Multiple Sorted Sets mode")
-        List(new RedisMultipleSortedSets(modeCache))
+        List(new RedisMultipleSortedSets(mode_PK_SS))
       }).flatten.toList
 
     require(writer.nonEmpty, s"No writers set for ${RedisConfigConstants.KCQL_CONFIG}!")

--- a/kafka-connect-redis/src/test/scala/com/datamountaineer/streamreactor/connect/redis/sink/writer/RedisMultipleSortedSetsTest.scala
+++ b/kafka-connect-redis/src/test/scala/com/datamountaineer/streamreactor/connect/redis/sink/writer/RedisMultipleSortedSetsTest.scala
@@ -16,6 +16,7 @@
 
 package com.datamountaineer.streamreactor.connect.redis.sink.writer
 
+import com.datamountaineer.streamreactor.connect.redis.sink.RedisSinkTask
 import com.datamountaineer.streamreactor.connect.redis.sink.config.{RedisConfig, RedisConfigConstants, RedisConnectionInfo, RedisSinkSettings}
 import org.apache.kafka.connect.data.{Schema, SchemaBuilder, Struct}
 import org.apache.kafka.connect.sink.SinkRecord
@@ -69,8 +70,6 @@ class RedisMultipleSortedSetsTest extends WordSpec with Matchers with BeforeAndA
       val sinkRecord3 = new SinkRecord(TOPIC, 0, null, null, schema, struct3, 3)
 
       val jedis = new Jedis(connectionInfo.host, connectionInfo.port)
-      // Clean up in-memory jedis
-      jedis.flushAll()
 
       writer.write(Seq(sinkRecord1))
       writer.write(Seq(sinkRecord2, sinkRecord3))
@@ -84,6 +83,62 @@ class RedisMultipleSortedSetsTest extends WordSpec with Matchers with BeforeAndA
 
     }
 
-  }
+    "multiple sorted sets task check" in {
+      val TOPIC = "sensorsTopic"
+      val KCQL = s"SELECT temperature, humidity FROM $TOPIC PK sensorID STOREAS SortedSet(score=ts)"
 
+      val props = Map(
+        RedisConfigConstants.REDIS_HOST->"localhost",
+        RedisConfigConstants.REDIS_PORT->"6379",
+        RedisConfigConstants.KCQL_CONFIG->KCQL
+      ).asJava
+
+      val task = new RedisSinkTask
+      task.start(props)
+
+      val schema = SchemaBuilder.struct().name("com.example.device")
+        .field("sensorID", Schema.STRING_SCHEMA)
+        .field("temperature", Schema.FLOAT64_SCHEMA)
+        .field("humidity", Schema.FLOAT64_SCHEMA)
+        .field("ts", Schema.INT64_SCHEMA).build()
+
+      val struct1 = new Struct(schema).put("sensorID", "sensor-sink-task").put("temperature", 60.4).put("humidity", 90.1).put("ts", 1482180657010L)
+      val sinkRecord1 = new SinkRecord(TOPIC, 0, null, null, schema, struct1, 1)
+
+      task.put(List(sinkRecord1).asJava)
+
+      val connectionInfo = new RedisConnectionInfo("localhost", 6379, None)
+      val jedis = new Jedis(connectionInfo.host, connectionInfo.port)
+      jedis.zcard("sensor-sink-task") shouldBe 1
+    }
+
+    "multiple sorted sets task check with prefix" in {
+      val TOPIC = "sensorsTopic"
+      val KCQL = s"INSERT INTO PREFIX- SELECT temperature, humidity FROM $TOPIC PK sensorID STOREAS SortedSet(score=ts)"
+
+      val props = Map(
+        RedisConfigConstants.REDIS_HOST->"localhost",
+        RedisConfigConstants.REDIS_PORT->"6379",
+        RedisConfigConstants.KCQL_CONFIG->KCQL
+      ).asJava
+
+      val task = new RedisSinkTask
+      task.start(props)
+
+      val schema = SchemaBuilder.struct().name("com.example.device")
+        .field("sensorID", Schema.STRING_SCHEMA)
+        .field("temperature", Schema.FLOAT64_SCHEMA)
+        .field("humidity", Schema.FLOAT64_SCHEMA)
+        .field("ts", Schema.INT64_SCHEMA).build()
+
+      val struct1 = new Struct(schema).put("sensorID", "sensor-sink-task-prefix").put("temperature", 60.4).put("humidity", 90.1).put("ts", 1482180657010L)
+      val sinkRecord1 = new SinkRecord(TOPIC, 0, null, null, schema, struct1, 1)
+
+      task.put(List(sinkRecord1).asJava)
+
+      val connectionInfo = new RedisConnectionInfo("localhost", 6379, None)
+      val jedis = new Jedis(connectionInfo.host, connectionInfo.port)
+      jedis.zcard("PREFIX-sensor-sink-task-prefix") shouldBe 1
+    }
+  }
 }

